### PR TITLE
Handle `Initialize` on subsystem in `Statevector`

### PR DESCRIPTION
**Pull request by Cryoris**
### Summary

This is a follow up to #8807 which provided a significant speedup if `Statevector` handled `Initialize` instructions -- however this PR didn't correctly account for initializations on a subpart of the qubits and returned a shrunken statevector! This PR fixes this and adds a test. 

### Details and comments

This causes a Qiskit ML tutorial to fail, thanks @manoelmarques for finding it! Also, since #8807 is not yet released, no reno is included (and no backport required 🙂).

### CODE REVIEW

File: qiskit/quantum_info/states/statevector.py [hunk 1]
```diff
@@ -862,20 +862,31 @@ def _evolve_instruction(statevec, obj, qargs=None):
...
-                statevec._data = Statevector.from_int(state, (2,) * obj.num_qubits)._data
+                initialization = Statevector.from_int(state, (2,) * obj.num_qubits)._data
             # state is initialized to the statevector
             else:
-                statevec._data = np.asarray(obj.params, dtype=complex)
+                initialization = np.asarray(obj.params, dtype=complex)
+
+            if qargs is None:
+                statevec._data = initialization
+            else:
```
Comment (ikkoham) [inline qiskit/quantum_info/states/statevector.py:882]: Line 865 assumes that quarts is None, but does the code reach this conditional branch?
(I could be misunderstanding something...)

File: qiskit/quantum_info/states/statevector.py [hunk 2]
```diff
@@ -862,20 +862,31 @@ def _evolve_instruction(statevec, obj, qargs=None):
...
             elif len(obj.params) == 1:
                 state = int(np.real(obj.params[0]))
-                statevec._data = Statevector.from_int(state, (2,) * obj.num_qubits)._data
+                initialization = Statevector.from_int(state, (2,) * obj.num_qubits)._data
             # state is initialized to the statevector
             else:
-                statevec._data = np.asarray(obj.params, dtype=complex)
+                initialization = np.asarray(obj.params, dtype=complex)
+
+            if qargs is None:
```
Comment (ElePT) [inline qiskit/quantum_info/states/statevector.py:880]: In a similar note to @ikkoham, if line 865 is true, then line 880 is always true, right?

File: qiskit/quantum_info/states/statevector.py [hunk 3]
```diff
@@ -862,20 +862,31 @@ def _evolve_instruction(statevec, obj, qargs=None):
             return statevec
         if isinstance(obj, Barrier):
             return statevec
-        if isinstance(obj, Initialize):
+        if isinstance(obj, Initialize) and qargs is None:
```
Comment (Cryoris) [inline qiskit/quantum_info/states/statevector.py:865]:
```suggestion
        if isinstance(obj, Initialize):
```

File: qiskit/quantum_info/states/statevector.py (hunk 2 above)
Comment (Cryoris) [reply qiskit/quantum_info/states/statevector.py:880]: Whoops, that check on line 865 was left from a test, that shouldn't be there!

File: qiskit/quantum_info/states/statevector.py (hunk 1 above)
Comment (Cryoris) [reply qiskit/quantum_info/states/statevector.py:882]: You're absolutely right -- that check on line 865 was left over from a test. I removed it!

**Pull request merged**